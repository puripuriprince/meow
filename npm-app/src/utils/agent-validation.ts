import { red, yellow } from 'picocolors'

import { websiteUrl } from '../config'
import { logger } from './logger'

/**
 * Validates agent definitions using the REST API
 * @param agentDefinitions The agent definitions to validate
 */
export async function validateAgentDefinitionsIfAuthenticated(
  agentDefinitions: any[],
): Promise<void> {
  // Only validate if there are agent configs
  if (!agentDefinitions || agentDefinitions.length === 0) {
    return
  }

  try {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    }

    const response = await fetch(`${websiteUrl}/api/agents/validate`, {
      method: 'POST',
      headers,
      body: JSON.stringify({ agentDefinitions }),
    })

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}))
      const errorMessage =
        (errorData as any).error ||
        `HTTP ${response.status}: ${response.statusText}`
      console.log(
        `\n${red('Agent Definition Validation Error:')} ${errorMessage}`,
      )
      return
    }

    const data = await response.json()

    if (data.validationErrors && data.validationErrors.length > 0) {
      const errorMessage = data.validationErrors
        .map((err: { filePath: string; message: string }) => err.message)
        .join('\n')
      console.log(
        `\n${yellow('Agent Definition Validation Warnings:')}\n${errorMessage}`,
      )
    }
  } catch (error) {
    logger.warn(
      {
        errorMessage: error instanceof Error ? error.message : String(error),
        errorStack: error instanceof Error ? error.stack : undefined,
      },
      'Failed to validate agent definitions via REST API',
    )
  }
}
