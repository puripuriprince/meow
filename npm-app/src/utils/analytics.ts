// Analytics disabled - no PostHog tracking
import type { AnalyticsEvent } from '@andji/common/constants/analytics-events'

// Store the identified user ID
let currentUserId: string | undefined

export let identified: boolean = false

export function initAnalytics() {
  // Analytics disabled - do nothing
  return
}

export async function flushAnalytics() {
  // No analytics to flush
  return
}

export function identifyUser(
  userId: string,
  properties?: Record<string, any>,
) {
  currentUserId = userId
  identified = true
  // No tracking
}

export function trackEvent(
  event: AnalyticsEvent,
  properties?: Record<string, any>,
) {
  // No tracking
}

export function logError(
  error: Error | unknown,
  context?: Record<string, any>,
) {
  // No error tracking
}