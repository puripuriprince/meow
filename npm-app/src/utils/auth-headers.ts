import { getUserCredentials } from '../credentials'
import { API_KEY_ENV_VAR } from '@andji/common/old-constants'

/**
 * Get the auth token from user credentials or environment variable
 */
export function getAuthToken(): string | undefined {
  const userCredentials = getUserCredentials()
  return userCredentials?.authToken || process.env[API_KEY_ENV_VAR]
}

/**
 * Create headers with x-andji-api-key for API requests
 */
export function createAuthHeaders(
  contentType = 'application/json',
): Record<string, string> {
  const headers: Record<string, string> = {
    'Content-Type': contentType,
  }

  const authToken = getAuthToken()
  if (authToken) {
    headers['x-andji-api-key'] = authToken
  }

  return headers
}

/**
 * Add x-andji-api-key to existing headers
 */
export function addAuthHeader(
  headers: Record<string, string>,
  authToken?: string,
): Record<string, string> {
  const token = authToken || getAuthToken()
  if (token) {
    return {
      ...headers,
      'x-andji-api-key': token,
    }
  }
  return headers
}
