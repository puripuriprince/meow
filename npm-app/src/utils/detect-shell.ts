import { execSync } from 'child_process'
import { platform } from 'os'

import { logger } from './logger'

export function detectShell():
  | 'bash'
  | 'zsh'
  | 'fish'
  | 'cmd.exe'
  | 'powershell'
  | 'unknown'
  | string {
  // Get the parent process environment
  const shell =
    process.env.SHELL || process.env.COMSPEC || process.env.PSModulePath

  try {
    // Handle Windows detection
    if (platform() === 'win32') {
      try {
        // Check if running in PowerShell
        execSync('$PSVersionTable', { stdio: 'pipe' })
        return 'powershell'
      } catch {
        // Check explicit CMD environment
        if (process.env.COMSPEC?.toLowerCase().includes('cmd.exe')) {
          return 'cmd.exe'
        }

        // Check parent process as final Windows check
        const parentProcess = execSync(
          'wmic process get ParentProcessId,CommandLine',
          { stdio: 'pipe' },
        )
          .toString()
          .toLowerCase()
        if (parentProcess.includes('powershell')) return 'powershell'
        if (parentProcess.includes('cmd.exe')) return 'cmd.exe'
      }
    }

    // Handle Unix-like systems
    if (shell) {
      const shellLower = shell.toLowerCase()
      if (shellLower.includes('bash')) return 'bash'
      if (shellLower.includes('zsh')) return 'zsh'
      if (shellLower.includes('fish')) return 'fish'
    }

    // Try to get the parent process name on Unix systems
    if (platform() !== 'win32') {
      const ppid = process.ppid
      const parentProcess = execSync(`ps -p ${ppid} -o comm=`, {
        stdio: 'pipe',
      })
        .toString()
        .trim()
      if (parentProcess) return parentProcess
    }
  } catch (error) {
    logger.error(
      {
        errorMessage: error instanceof Error ? error.message : String(error),
        errorStack: error instanceof Error ? error.stack : undefined,
        platform: platform(),
        shell,
      },
      'Error detecting shell',
    )
  }

  return 'unknown'
}
