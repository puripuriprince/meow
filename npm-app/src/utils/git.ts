import { execFileSync } from 'child_process'
import fs, { existsSync, statSync } from 'fs'
import path from 'path'

import { isValidProjectRoot } from '@andji/common/util/file'
import gitUrlParse from 'git-url-parse'

import { getWorkingDirectory } from '../project-files'
import { logger } from './logger'

// Dynamic import for isomorphic-git
async function getIsomorphicGit() {
  const git = await import('isomorphic-git')
  return git
}

/**
 * Checks if the native git command is available on the system.
 * Caches the result to avoid repeated checks.
 * @returns boolean indicating if git command is available
 */
let cachedGitAvailable: boolean | null = null
export function gitCommandIsAvailable(): boolean {
  if (cachedGitAvailable === null) {
    try {
      execFileSync('git', ['--version'], { stdio: 'ignore' })
      cachedGitAvailable = true
    } catch (error) {
      logger.error(
        {
          errorMessage: error instanceof Error ? error.message : String(error),
          errorStack: error instanceof Error ? error.stack : undefined,
        },
        'Git command not available',
      )
      cachedGitAvailable = false
    }
  }

  return cachedGitAvailable
}

export function findGitRoot(startDir: string): string | null {
  let currentDir = startDir

  while (currentDir !== path.parse(currentDir).root) {
    const gitDir = path.join(currentDir, '.git')
    if (existsSync(gitDir) && statSync(gitDir).isDirectory()) {
      return isValidProjectRoot(currentDir) ? currentDir : null
    }
    currentDir = path.dirname(currentDir)
  }

  return null
}

function parseRepoName(remoteUrl: string | undefined): string | undefined {
  if (!remoteUrl) return undefined
  try {
    const { full_name, owner, name } = gitUrlParse(remoteUrl)
    return full_name || (owner && name ? `${owner}/${name}` : undefined)
  } catch {
    return undefined
  }
}

// Always uses isomorphic-git since it can more easily run asynchronously
export async function getRepoMetrics(providedRemoteUrl?: string): Promise<{
  ageDays?: number
  trackedFiles?: number
  commits?: number
  repoName?: string
  repoUrl?: string
  owner?: string
  repo?: string
  commitsLast30Days?: number
  authorsLast30Days?: number
}> {
  const t = Date.now()
  const cwd = getWorkingDirectory()
  const root = findGitRoot(cwd)

  if (!root) {
    return {}
  }

  const gitDir = path.join(root, '.git')
  const { log, listFiles, getConfig } = await getIsomorphicGit()

  const commitsArr = await log({ fs, dir: root, gitdir: gitDir }).catch(
    (error) => {
      logger.error(
        { error },
        'Error fetching git log. Is this an empty git repo?',
      )
      return [] as any[]
    },
  )
  const firstCommit = commitsArr.at(-1) // earliest

  const tracked = await listFiles({ fs, dir: root, gitdir: gitDir })

  // Use provided remote URL or fetch from git config
  let remoteUrl: string | undefined
  if (providedRemoteUrl) {
    remoteUrl = providedRemoteUrl
  } else {
    remoteUrl = await getConfig({
      fs,
      gitdir: gitDir,
      path: 'remote.origin.url',
    })
  }

  // Parse owner and repo from the remote URL
  let owner: string | undefined
  let repo: string | undefined
  if (remoteUrl) {
    try {
      const parsed = gitUrlParse(remoteUrl)
      owner = parsed.owner
      repo = parsed.name
    } catch (error) {
      // If parsing fails, owner and repo will remain undefined
    }
  }

  const nowSec = Math.floor(Date.now() / 1000)
  const THIRTY_DAYS = 30 * 24 * 60 * 60
  const recent = commitsArr.filter(
    (c) => nowSec - c.commit.committer.timestamp <= THIRTY_DAYS,
  )
  const authors = new Set(
    recent.map(
      (c) => `${c.commit.author.name}|${c.commit.author.email?.toLowerCase()}`,
    ),
  )

  const res = {
    ageDays: firstCommit
      ? Math.floor(
          (Date.now() / 1000 - firstCommit.commit.committer.timestamp) / 86_400,
        )
      : 0,
    trackedFiles: tracked.length,
    commits: commitsArr.length,
    repoName: parseRepoName(remoteUrl ?? undefined),
    repoUrl: remoteUrl ?? undefined,
    owner,
    repo,
    commitsLast30Days: recent.length,
    authorsLast30Days: authors.size,
  }

  return res
}
