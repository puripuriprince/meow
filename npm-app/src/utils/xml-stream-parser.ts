import {
  endsAgentStepParam,
  toolNameParam,
  toolXmlName,
} from '@andji/common/tools/constants'
import { getPartialJsonDelta } from '@andji/common/util/partial-json-delta'
import { Saxy } from '@andji/common/util/saxy'

import { defaultToolCallRenderer } from './tool-renderers'

import type { ToolCallRenderer } from './tool-renderers'
import { MarkdownStreamRenderer } from '../display/markdown-renderer'

/**
 * Creates a transform stream that processes XML tool calls
 * @param renderer Custom renderer for tool calls or a map of renderers per tool
 * @param callback Optional callback function to receive processed chunks
 * @returns Transform stream
 */
export function createXMLStreamParser(
  renderer: Record<string, ToolCallRenderer>,
  callback?: (chunk: string) => void,
) {
  // Create parser with tool schema validation
  const parser = new Saxy({ [toolXmlName]: [] })

  const md = new MarkdownStreamRenderer({
    width: process.stdout.columns || 80,
    isTTY: process.stdout.isTTY,
    syntaxHighlight: true,
    streamingMode: 'smart', // Use smart content-aware streaming with loading indicators
  })

  // Current state
  let inToolCallTag = false
  let currentTool: string | null = null
  let params: Record<string, string> = {}
  let completedParams: string[] = []
  let currentParam: string | null = null
  let paramsContent = ''

  // Helper to get the appropriate renderer for the current tool
  const getRenderer = (toolName: string): ToolCallRenderer => {
    if (!renderer) return defaultToolCallRenderer

    // If renderer is a map of tool-specific renderers
    if (typeof renderer === 'object' && !('onToolStart' in renderer)) {
      return (
        (renderer as Record<string, ToolCallRenderer>)[toolName] ||
        defaultToolCallRenderer
      )
    }

    // If renderer is a single renderer
    return renderer as ToolCallRenderer
  }

  // Set up event handlers
  parser.on('tagopen', (tag) => {
    inToolCallTag = true
  })

  parser.on('text', (data) => {
    if (!inToolCallTag) {
      const outs = md.write(data.contents)
      for (const out of outs) {
        parser.push(out)
        if (callback) callback(out)
      }
      return
    }

    const prevParamsContent = paramsContent
    paramsContent += data.contents
    const {
      delta,
      result,
      lastParam: { key: lastKey, complete: lastComplete },
    } = getPartialJsonDelta(paramsContent, prevParamsContent)
    if (
      lastKey === toolNameParam &&
      lastComplete &&
      delta[lastKey] === undefined
    ) {
      delta[lastKey] = ''
    }
    for (const [key, value] of Object.entries(delta)) {
      if (key === toolNameParam) {
        if (key === lastKey && !lastComplete) {
          continue
        }
        if (currentTool !== null) {
          continue
        }
        // start tool
        const toolRenderer = getRenderer(result[key])
        if (toolRenderer.onToolStart) {
          const output = toolRenderer.onToolStart(result[key], {})
          if (typeof output === 'string') {
            parser.push(output)
            if (callback) callback(output)
          } else if (output !== null) {
            output()
          }
        }
        currentTool = result[key]
        continue
      }

      // handle tool params
      const stringValue =
        typeof value === 'string' ? value : JSON.stringify(value)
      if (key === endsAgentStepParam) {
        continue
      }
      const toolName = result[toolNameParam]
      if (currentParam !== null && currentParam !== key) {
        const toolRenderer = getRenderer(toolName)
        if (toolRenderer.onParamEnd) {
          const output = toolRenderer.onParamEnd(
            currentParam,
            toolName,
            result[currentParam],
          )
          if (typeof output === 'string') {
            parser.push(output)
            if (callback) callback(output)
          } else if (output !== null) {
            output()
          }
        }
        completedParams.push(currentParam)
      }
      if (completedParams.includes(key)) {
        currentParam = null
        continue
      }
      currentParam = key
      const toolRenderer = getRenderer(toolName)
      if (params[key] === undefined) {
        if (toolRenderer.onParamStart) {
          const output = toolRenderer.onParamStart(key, toolName)
          if (typeof output === 'string') {
            parser.push(output)
            if (callback) callback(output)
          } else if (output !== null) {
            output()
          }
        }
      }
      if (toolRenderer.onParamChunk) {
        const output = toolRenderer.onParamChunk(stringValue, key, toolName)
        if (typeof output === 'string') {
          parser.push(output)
          if (callback) callback(output)
        } else if (output !== null) {
          output()
        }
      }
      if (key === lastKey && lastComplete) {
        const toolRenderer = getRenderer(toolName)
        if (toolRenderer.onParamEnd) {
          const output = toolRenderer.onParamEnd(
            key,
            toolName,
            typeof result[key] === 'string'
              ? result[key]
              : JSON.stringify(result[key]),
          )
          if (typeof output === 'string') {
            parser.push(output)
            if (callback) callback(output)
          } else if (output !== null) {
            output()
          }
        }
        completedParams.push(key)
        currentParam = null
      }
    }
    params = Object.fromEntries(
      Object.entries(result).map(([k, v]) => [
        k,
        typeof v === 'string' ? v : JSON.stringify(v),
      ]),
    )
  })

  parser.on('tagclose', (tag) => {
    if (!inToolCallTag) {
      return
    }
    const toolName = params[toolNameParam]
    const toolRenderer = getRenderer(toolName)

    if (currentParam !== null) {
      if (toolRenderer.onParamEnd) {
        const output = toolRenderer.onParamEnd(
          currentParam,
          toolName,
          params[currentParam],
        )
        if (typeof output === 'string') {
          parser.push(output)
          if (callback) callback(output)
        } else if (output !== null) {
          output()
        }
      }
    }

    if (toolRenderer.onToolEnd) {
      const output = toolRenderer.onToolEnd(toolName, params)
      if (typeof output === 'string') {
        parser.push(output)
        if (callback) callback(output)
      } else if (output !== null) {
        output()
      }
    }

    inToolCallTag = false
    paramsContent = ''
    params = {}
    currentParam = null
    currentTool = null
    completedParams = []
  })

  parser._flush = function (done: (error?: Error | null) => void) {
    const rem = md.end()
    if (rem) {
      this.push(rem)
      if (callback) callback(rem)
    }
    done()
  }

  // Override destroy to ensure markdown renderer cleanup
  const originalDestroy = parser.destroy.bind(parser)
  parser.destroy = function (error?: Error) {
    md.cleanup()
    return originalDestroy(error)
  }

  return parser
}
