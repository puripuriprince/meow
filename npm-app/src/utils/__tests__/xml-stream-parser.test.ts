import fs from 'fs'
import path from 'path'
import { Writable } from 'stream'

import { getToolCallString } from '@andji/common/tools/utils'
import { describe, expect, test } from 'bun:test'
import stripAnsi from 'strip-ansi'

import { defaultToolCallRenderer } from '../tool-renderers'
import { createXMLStreamParser } from '../xml-stream-parser'

const toolRenderers = {
  run_terminal_command: defaultToolCallRenderer,
  write_file: defaultToolCallRenderer,
  read_files: defaultToolCallRenderer,
}

describe('createXMLStreamParser', () => {
  test('pipes output to writable stream', async () => {
    const xml = getToolCallString('run_terminal_command', {
      command: 'echo hello',
    })
    let result = ''

    const processor = createXMLStreamParser(toolRenderers)

    const writable = new Writable({
      write(chunk, encoding, callback) {
        result += chunk.toString()
        callback()
      },
    })

    processor.pipe(writable)
    processor.write(xml)
    processor.end()

    // Wait for the stream to finish
    await new Promise<void>((resolve) => {
      writable.on('finish', resolve)
    })

    expect(result).toContain('Run Terminal Command')
    expect(result).toContain('echo hello')
  })

  test('handles multiple tool calls in sequence', async () => {
    const xml =
      getToolCallString('run_terminal_command', { command: 'ls' }) +
      getToolCallString('read_files', { paths: ['file.txt'] })
    let result = ''

    const processor = createXMLStreamParser(toolRenderers)

    const writable = new Writable({
      write(chunk, encoding, callback) {
        result += chunk.toString()
        callback()
      },
    })

    processor.pipe(writable)
    processor.write(xml)
    processor.end()

    // Wait for the stream to finish
    await new Promise<void>((resolve) => {
      writable.on('finish', resolve)
    })

    expect(result).toContain('Run Terminal Command')
    expect(result).toContain('ls')
    expect(result).toContain('Read Files')
    expect(result).toContain('file.txt')
  })

  test('handles text content between tool calls', async () => {
    const xml =
      getToolCallString('run_terminal_command', { command: 'ls' }) +
      'Some text between tool calls' +
      getToolCallString('read_files', { paths: ['file.txt'] })
    let result = ''

    const processor = createXMLStreamParser(toolRenderers)

    const writable = new Writable({
      write(chunk, encoding, callback) {
        result += chunk.toString()
        callback()
      },
    })

    processor.pipe(writable)
    processor.write(xml)
    processor.end()

    // Wait for the stream to finish
    await new Promise<void>((resolve) => {
      writable.on('finish', resolve)
    })

    expect(result).toContain('Run Terminal Command')
    expect(result).toContain('ls')
    expect(result).toContain('Some text between tool calls')
    expect(result).toContain('Read Files')
    expect(result).toContain('file.txt')
  })

  test('processes chunks incrementally with output verification', async () => {
    // Define chunks that will be written one at a time
    const terminalCommand = getToolCallString('run_terminal_command', {
      command: 'npm install',
    })
    const codeSearch = getToolCallString('code_search', { pattern: 'function' })

    // Split the XML into chunks
    const chunks = [
      terminalCommand.slice(0, 15),
      terminalCommand.slice(15, 30),
      terminalCommand.slice(30, 40),
      terminalCommand.slice(40, 50),
      terminalCommand.slice(50),
      codeSearch.slice(0, 15),
      codeSearch.slice(15),
    ]

    // Array to store output chunks as they're produced
    const outputChunks: string[] = []

    // Create a writable stream that captures each output chunk
    const writable = new Writable({
      write(chunk, encoding, callback) {
        const chunkStr = chunk.toString()
        outputChunks.push(chunkStr)
        callback()
      },
    })

    // Create the processor
    const processor = createXMLStreamParser(toolRenderers)
    processor.pipe(writable)

    // Process each chunk with a small delay between them
    for (const chunk of chunks) {
      processor.write(chunk)
      // Small delay to simulate real-world streaming
      await new Promise((resolve) => setTimeout(resolve, 5))
    }
    processor.end()

    // Wait for the stream to finish
    await new Promise<void>((resolve) => {
      writable.on('finish', resolve)
    })

    // Verify the complete output
    const fullOutput = stripAnsi(outputChunks.join(''))
    expect(fullOutput).toContain('Run Terminal Command')
    expect(fullOutput).toContain('npm install')
    expect(fullOutput).toContain('Code Search')
    expect(fullOutput).toContain('function')

    // Verify incremental output - but don't make assumptions about which chunks contain what
    // Just check that we got multiple output chunks (incremental processing)
    expect(outputChunks.length).toBeGreaterThan(1)

    // Check that the terminal command appears before the code search in the output
    const terminalCommandIndex = fullOutput.indexOf('Run Terminal Command')
    const codeSearchIndex = fullOutput.indexOf('Code Search')
    expect(terminalCommandIndex).toBeLessThan(codeSearchIndex)
  })

  test('processes text content incrementally', async () => {
    // Define chunks that will be written one at a time
    const chunks = ['hi', 'hi2', 'hi3 <yo>\nyo\n</yo>']

    // Array to store output chunks as they're produced
    const outputChunks: string[] = []

    // Create a writable stream that captures each output chunk
    const writable = new Writable({
      write(chunk, encoding, callback) {
        const chunkStr = chunk.toString()
        outputChunks.push(chunkStr)
        callback()
      },
    })

    // Create the processor
    const processor = createXMLStreamParser(toolRenderers)
    processor.pipe(writable)

    // Process each chunk with a small delay between them
    for (const chunk of chunks) {
      processor.write(chunk)
      // Small delay to simulate real-world streaming
      await new Promise((resolve) => setTimeout(resolve, 5))
    }
    processor.end()

    // Wait for the stream to finish
    await new Promise<void>((resolve) => {
      writable.on('finish', resolve)
    })

    // Verify the complete output
    const fullOutput = outputChunks.join('')
    expect(fullOutput).toEqual('hihi2hi3 <yo>\nyo\n</yo>')

    expect(outputChunks.length).toBeGreaterThanOrEqual(2)
  })

  test('handles xml entities in chunks correctly', async () => {
    // This test demonstrates an issue where chunks containing & are delayed
    const chunks = ['start', ' <tag>before & after</tag> ', 'end']

    const outputChunks: string[] = []

    const writable = new Writable({
      write(chunk, encoding, callback) {
        const chunkStr = chunk.toString()
        outputChunks.push(chunkStr)
        callback()
      },
    })

    const processor = createXMLStreamParser(toolRenderers)
    processor.pipe(writable)

    // Process each chunk with a small delay
    for (const chunk of chunks) {
      processor.write(chunk)
      await new Promise((resolve) => setTimeout(resolve, 5))
    }
    processor.end()

    // Wait for stream to finish
    await new Promise<void>((resolve) => {
      writable.on('finish', resolve)
    })

    // The test should fail because the chunk containing & is delayed
    expect(outputChunks.join('')).toEqual(chunks.join(''))
  })

  test('preserves spaces between words in plain text', async () => {
    // Test that spaces are preserved when text is split across chunks
    const chunks = [
      'andji is an AI-powered terminal-based code editor that helps',
      ' developers write code through natural language instructions.',
    ]

    const outputChunks: string[] = []

    const writable = new Writable({
      write(chunk, encoding, callback) {
        const chunkStr = chunk.toString()
        outputChunks.push(chunkStr)
        callback()
      },
    })

    const processor = createXMLStreamParser(toolRenderers)
    processor.pipe(writable)

    // Process each chunk
    for (const chunk of chunks) {
      processor.write(chunk)
    }
    processor.end()

    // Wait for stream to finish
    await new Promise<void>((resolve) => {
      writable.on('finish', resolve)
    })

    const fullOutput = outputChunks.join('')
    // Should contain "helps developers" with the space preserved
    expect(fullOutput).toContain('helps developers')
    expect(fullOutput).toContain(
      'andji is an AI-powered terminal-based code editor that helps developers',
    )
  })

  test('preserves spaces in complex sentence', async () => {
    // Test the exact scenario the user mentioned
    const chunks = [
      'The goal is to make expert',
      ' software engineers move even faster by reducing time on common programming tasks.',
    ]

    const outputChunks: string[] = []

    const writable = new Writable({
      write(chunk, encoding, callback) {
        const chunkStr = chunk.toString()
        outputChunks.push(chunkStr)
        callback()
      },
    })

    const processor = createXMLStreamParser(toolRenderers)
    processor.pipe(writable)

    // Process each chunk
    for (const chunk of chunks) {
      processor.write(chunk)
    }
    processor.end()

    // Wait for stream to finish
    await new Promise<void>((resolve) => {
      writable.on('finish', resolve)
    })

    const fullOutput = outputChunks.join('')
    // The space between "expert" and "software" should be preserved
    expect(fullOutput).not.toContain('expertsoftware')
    expect(fullOutput).toContain('expert software')
    expect(fullOutput).toContain(
      'The goal is to make expert software engineers',
    )
  })

  test('real world example write to 4 files', async () => {
    // Read the file content directly - no need for complex unescaping
    const response = fs.readFileSync(
      path.join(__dirname, './response-example-4-files.txt'),
      'utf-8',
    )

    // Array to store output chunks as they're produced
    const outputChunks: string[] = []

    // Create a writable stream that captures each output chunk
    const writable = new Writable({
      write(chunk, encoding, callback) {
        const chunkStr = chunk.toString()
        outputChunks.push(chunkStr)
        callback()
      },
    })

    let writeFileStartCount = 0
    let writeFileEndCount = 0
    const testToolRenderers = {
      ...toolRenderers,
      write_file: {
        ...defaultToolCallRenderer,
        onToolStart: (toolName: string) => {
          writeFileStartCount++
          return 'Write File'
        },
        onParamStart: (paramName: string, toolName: string) => {
          return null
        },
        onParamEnd: (paramName: string, toolName: string, content: string) => {
          return null
        },
        onToolEnd: (toolName: string, params: Record<string, string>) => {
          writeFileEndCount++
          return null
        },
      },
    }

    // Create the processor
    const processor = createXMLStreamParser(testToolRenderers)
    processor.pipe(writable)

    processor.write(response)
    processor.end()

    // Wait for the stream to finish
    await new Promise<void>((resolve) => {
      writable.on('finish', resolve)
    })

    // Verify the complete output
    const fullOutput = outputChunks.join('')
    expect(fullOutput).toContain('Write File') // Check for expected output without worrying about formatting
    expect(writeFileEndCount).toBe(4)
    // We expect multiple chunks for a large response
    expect(outputChunks.length).toBeGreaterThan(1)
  })
})
