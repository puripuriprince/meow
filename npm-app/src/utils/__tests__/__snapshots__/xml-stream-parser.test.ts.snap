// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getBackgroundProcessInfoString formats a running process correctly 1`] = `
"<background_process>
<process_id>123</process_id>
<command>npm test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>2000</duration_ms>
<stdout>test output</stdout>
<stderr>test error</stderr>
<status>running</status>
</background_process>"
`;

exports[`getBackgroundProcessInfoString formats a completed process correctly 1`] = `
"<background_process>
<process_id>456</process_id>
<command>npm build</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>1000</duration_ms>
<stdout>build successful</stdout>
<status>completed</status>
<exit_code>0</exit_code>
</background_process>"
`;

exports[`getBackgroundProcessInfoString formats an errored process correctly 1`] = `
"<background_process>
<process_id>789</process_id>
<command>invalid-command</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>1500</duration_ms>
<stderr>command not found</stderr>
<status>error</status>
<exit_code>1</exit_code>
<signal_code>SIGTERM</signal_code>
</background_process>"
`;

exports[`getBackgroundProcessInfoString handles new output since last report 1`] = `
"<background_process>
<process_id>102</process_id>
<command>echo test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>1000</duration_ms>
<stdout>[PREVIOUS OUTPUT]
 more output</stdout>
<status>completed</status>
</background_process>"
`;

exports[`getBackgroundProcessInfoString handles no new content 1`] = `
"<background_process>
<process_id>103</process_id>
<command>echo test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>1000</duration_ms>
<status>running</status>
</background_process>"
`;

exports[`getBackgroundProcessInfoString handles new stderr without when no previous stderr 1`] = `
"<background_process>
<process_id>104</process_id>
<command>echo test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>1000</duration_ms>
<stderr>new error</stderr>
<status>error</status>
</background_process>"
`;

exports[`getBackgroundProcessInfoString handles new stdout without when no previous stdout 1`] = `
"<background_process>
<process_id>105</process_id>
<command>echo test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>2000</duration_ms>
<stdout>first output</stdout>
<status>running</status>
</background_process>"
`;

exports[`getBackgroundProcessInfoString reports completed process with new stderr even if stdout unchanged 1`] = `
"<background_process>
<process_id>106</process_id>
<command>echo test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>1000</duration_ms>
<stderr>new error</stderr>
<status>completed</status>
</background_process>"
`;

exports[`getBackgroundProcessInfoString reports completed process with new stdout even if stderr unchanged 1`] = `
"<background_process>
<process_id>107</process_id>
<command>echo test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>1000</duration_ms>
<stdout>[PREVIOUS OUTPUT]
 more</stdout>
<status>completed</status>
</background_process>"
`;

exports[`getBackgroundProcessInfoString reports process when status changes even without output changes 1`] = `
"<background_process>
<process_id>108</process_id>
<command>echo test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>1000</duration_ms>
<status>completed</status>
</background_process>"
`;

exports[`getBackgroundProcessInfoString calculates duration from endTime when available 1`] = `
"<background_process>
<process_id>109</process_id>
<command>echo test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>1500</duration_ms>
<stdout>test</stdout>
<status>completed</status>
</background_process>"
`;

exports[`getBackgroundProcessInfoString calculates duration from current time when no endTime 1`] = `
"<background_process>
<process_id>110</process_id>
<command>echo test</command>
<start_time_utc>1970-01-01T00:00:01.000Z</start_time_utc>
<duration_ms>2000</duration_ms>
<stdout>test</stdout>
<status>running</status>
</background_process>"
`;
